### PR TITLE
Rewrite Versioned Shared Library logic and implement SOVERSION logic from PR 3753

### PR DESCRIPTION
**Pull request by bdbaddog**
The logic has been simplified and is now hopefully easier to understand, maintain, and enhance.

Some details From #3753 

The SOVERSION is a new environment variable which works very similar to the SONAME environment variable, except instead of providing the full library name as you do with SONAME, you just provide an number which will generally be the major version number in the symlinked library.

The SHLIBVERSION environment variable is also very similar to this, except that the SHLIBVERSION is generally the specific version number of the library, and used on the actual output library, instead of the symlinked library.

Please review https://tldp.org/HOWTO/Program-Library-HOWTO/shared-libraries.html for more information on the standard conventions for library versioning in linux.

Also note that macOS uses a similar convention for versioning libraries, so this PR also adds the variable to apple linker.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by bdbaddog** [comment]
@dmoody256 - please take a look when you get a chance.

**Comment by dmoody256** [comment]
@bdbaddog  looks good, much less complex and more organized. Looks like windows is broken, which I am assuming it just needs a bit of tweaking to fit with the new common link? Anyways the current common link structure is much better.

File: SCons/Defaults.py [hunk 1]
```diff
@@ -542,43 +572,69 @@ def __call__(self, *args, **kw):
             frame = frame.f_back
         return None
 
-# if $version_var is not empty, returns env[flags_var], otherwise returns None
+
 def __libversionflags(env, version_var, flags_var):
+    """
+    if $version_var is not empty, returns env[flags_var], otherwise returns None
```
Comment (mwichmann) [inline SCons/Defaults.py]: just asking - is None a suitable return value here (I realize it already was that) - applies to new bit below also.

File: SCons/Platform/aix.py [hunk 2]
```diff
@@ -71,6 +71,7 @@ def generate(env):
     posix.generate(env)
     #Based on AIX 5.2: ARG_MAX=24576 - 3000 for environment expansion
     env['MAXLINELENGTH']  = 21576
+    env['SHLIBSUFFIX'] = '.a'
```
Comment (mwichmann) [inline SCons/Platform/aix.py:74]: AIX uses `.a` for shared libs?

File: SCons/Tool/FortranCommon.py [hunk 3]
```diff
@@ -4,8 +4,6 @@
 
 """
 
-#
-# __COPYRIGHT__
 #
```
Comment (mwichmann) [inline SCons/Tool/FortranCommon.py:10]: insert current standard header bit?  (MIT License \n Copyright SCons ... etc.)

File: SCons/Tool/applelink.py [hunk 4]
```diff
@@ -212,10 +165,21 @@ def generate(env):
     # pre/suffixes:
     env['LDMODULEPREFIX'] = ''
     env['LDMODULESUFFIX'] = ''
-    env['LDMODULEFLAGS'] = SCons.Util.CLVar('$LINKFLAGS -bundle')
-    env['LDMODULECOM'] = '$LDMODULE -o ${TARGET} $LDMODULEFLAGS $SOURCES $_LIBDIRFLAGS $_LIBFLAGS $_FRAMEWORKPATH $_FRAMEWORKS $FRAMEWORKSFLAGS'
-
-    env['__SHLIBVERSIONFLAGS'] = '${__libversionflags(__env__,"SHLIBVERSION","_SHLIBVERSIONFLAGS")}'
+    env['LDMODULEFLAGS'] = CLVar('$LINKFLAGS -bundle')
+    env[
```
Comment (mwichmann) [inline SCons/Tool/applelink.py]: that's an awkward looking break after the opening bracket. Plus, string pasting works without needing the backslash

**Review by mwichmann — COMMENTED** [review]
Looks good so far.  These comments are just nitpicky things.

**Comment by bdbaddog** [comment]
> @bdbaddog looks good, much less complex and more organized. Looks like windows is broken, which I am assuming it just needs a bit of tweaking to fit with the new common link? Anyways the current common link structure is much better.

Yup. WIP.. I was developing on mac, then tested on linux after CI, sorted that, Windows is next.

File: SCons/Platform/aix.py (hunk 2 above)
Comment (bdbaddog) [reply SCons/Platform/aix.py:74]: Yup.. AIX is a strange beastie..

File: SCons/Tool/applelink.py (hunk 4 above)
Comment (bdbaddog) [reply SCons/Tool/applelink.py]: Autoformatter did that..

File: SCons/Tool/FortranCommon.py (hunk 3 above)
Comment (bdbaddog) [reply SCons/Tool/FortranCommon.py:10]: done.

File: SCons/Tool/applelink.py (hunk 4 above)
Comment (bdbaddog) [reply SCons/Tool/applelink.py]: fixed.

**Comment by bdbaddog** [comment]
> @bdbaddog looks good, much less complex and more organized. Looks like windows is broken, which I am assuming it just needs a bit of tweaking to fit with the new common link? Anyways the current common link structure is much better.

Yes. Just pushed a fix.

That leaves a bit of work to resolve the various D compilers usage of versioned shared libraries.

**Comment by acmorrow** [comment]
@bdbaddog - A question about this, having not really been keeping up with it.  After this change, do the various symlinks come back as indexable targets from the `SharedLibrary` builder call? They didn't used to, but it would be nicer if they did.

File: SCons/Defaults.py [hunk 5]
```diff
@@ -127,6 +135,7 @@ def SharedFlagChecker(source, target, env):
 # the rest of those in Tool/__init__.py, but I'm not sure where else
 # they should go.  Leave them here for now.
 import SCons.Scanner.Dir
```
[the remark this answers is not in the document]
Comment (bdbaddog) [reply SCons/Defaults.py]: Yup.

**Comment by bdbaddog** [comment]
> @bdbaddog - A question about this, having not really been keeping up with it. After this change, do the various symlinks come back as indexable targets from the `SharedLibrary` builder call? They didn't used to, but it would be nicer if they did.

@acmorrow 
No change to that in this PR.

It does make sense that they would be in the list of targets, but I'd like to avoid changing any existing functionality in this PR, just add SOVERSION and ensure nothing is broken with the rewrite of the implementation.

**Comment by acmorrow** [comment]
@bdbaddog - No problem, and understood about limiting the scope. Should I file a ticket requesting that change?

**Comment by bdbaddog** [comment]
@acmorrow - double check there's not already one? I think @dragon512 asked about this a while back as well.

From some previous discussion it might break internal usage and/or external expectations so we'd probably have to roll a new feature version number instead of a patch. Which might not be a bad idea for this change anyway..

**Comment by bdbaddog** [comment]
@russel - Any reason D needs a separate versioned shared library env var?  Can we not just use SHLIBVERSION as we do for all other shared library types? Same for SONAME (and SOVERSION which is new)

**Pull request merged**